coord/src/rmr/rmr.c

```diff
@@ -445,25 +372,145 @@ size_t MR_NumHosts() {
 
 void SetMyPartition(MRClusterTopology *ct, MRClusterShard *myShard);
 /* on-loop update topology request. This can't be done from the main thread */
-static void uvUpdateTopologyRequest(struct MRRequestCtx *mc) {
-  MRCLuster_UpdateTopology(cluster_g, (MRClusterTopology *)mc->ctx);
-  if (cluster_g->myshard) {
-    SetMyPartition((MRClusterTopology *)mc->ctx, cluster_g->myshard);
+static void uvUpdateTopologyRequest(void *p) {
+  MRClusterTopology *topo = p;
+  MRCLuster_UpdateTopology(cluster_g, topo);
+  if (cluster_g->myShard) {
+    SetMyPartition(topo, cluster_g->myShard);
   }
-  RQ_Done(rq_g);
-  // fprintf(stderr, "topo update: conc requests: %d\n", concurrentRequests_g);
-  rm_free(mc);
 }
 
 /* Set a new topology for the cluster */
 void MR_UpdateTopology(MRClusterTopology *newTopo) {
   // enqueue a request on the io thread, this can't be done from the main thread
-  struct MRRequestCtx *rc = rm_calloc(1, sizeof(*rc));
-  rc->ctx = newTopo;
-  rc->cb = uvUpdateTopologyRequest;
-  rc->protocol = 0;
+  RQ_Push_Topology(uvUpdateTopologyRequest, newTopo);
+}
+
+static void uvReplyClusterInfo(void *p) {
+  RedisModuleBlockedClient *bc = p;
+  RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(bc);
+  MR_ReplyClusterInfo(ctx, cluster_g->topo);
+  RedisModule_FreeThreadSafeContext(ctx);
+  RedisModule_BlockedClientMeasureTimeEnd(bc);
+  RedisModule_UnblockClient(bc, NULL);
+}
+
+void MR_uvReplyClusterInfo(RedisModuleCtx *ctx) {
+  RedisModuleBlockedClient *bc = RedisModule_BlockClient(ctx, NULL, NULL, NULL, 0);
+  RedisModule_BlockedClientMeasureTimeStart(bc);
+  RQ_Push(rq_g, uvReplyClusterInfo, bc);
+}
+
+void MR_ReplyClusterInfo(RedisModuleCtx *ctx, MRClusterTopology *topo) {
+  RS_AutoMemory(ctx);
+  RedisModule_Reply _reply = RedisModule_NewReply(ctx), *reply = &_reply;
+
+  const char *hash_func_str;
+  switch (topo ? topo->hashFunc : MRHashFunc_None) {
+  case MRHashFunc_CRC12:
+    hash_func_str = MRHASHFUNC_CRC12_STR;
+    break;
+  case MRHashFunc_CRC16:
+    hash_func_str = MRHASHFUNC_CRC16_STR;
+    break;
+  default:
+    hash_func_str = "n/a";
+    break;
+  }
+  const char *cluster_type_str = clusterConfig.type == ClusterType_RedisOSS ? CLUSTER_TYPE_OSS : CLUSTER_TYPE_RLABS;
+  size_t partitions = topo ? topo->numShards : 0;
+
+  //-------------------------------------------------------------------------------------------
+  if (reply->resp3) { // RESP3 variant
```

Comment:
Whats the different between resp3 and resp2? If its only the reply with map/array then Redis will handle it for us.

Reply:
I just moved this function from `coord/src/module.c` to here. I think there are some minor changes (like naming the first and last slots). Not something I want to break now. but we can unify it